File: TelerikUIExamples/TelerikUIExamples/CalendarTransitonEffects.m
//
//  CalendarTransitionEffects.m
//  TelerikUIExamples
//
//  Copyright (c) 2014 Telerik. All rights reserved.
//

#import "CalendarTransitionEffects.h"
#import <TelerikUI/TelerikUI.h>

@interface CalendarTransitionEffects () <TKCalendarPresenterDelegate, TKCalendarDelegate>

@property (nonatomic, strong) TKCalendar *calendarView;
@property (nonatomic, strong) UIButton *buttonPrev;
@property (nonatomic, strong) UIButton *buttonNext;

@end

@implementation CalendarTransitionEffects
{
    NSInteger _colorIndex;
    NSInteger _oldColorIndex;
    NSArray *_colors;
    TKCalendarTransitionMode _transitionMode;
    UIView *_contentView;
}

- (instancetype)initWithNibName:(NSString *)nibNameOrNil bundle:(NSBundle *)nibBundleOrNil
{
    self = [super initWithNibName:nibNameOrNil bundle:nibBundleOrNil];
    if (self) {
        _colorIndex = 0;
        _colors = @[ [UIColor colorWithRed:0. green:1. blue:0. alpha:0.3],
                     [UIColor colorWithRed:1. green:0. blue:0. alpha:0.3],
                     [UIColor colorWithRed:0. green:0. blue:1. alpha:0.3],
                     [UIColor colorWithRed:1. green:1. blue:0. alpha:0.3],
                     [UIColor colorWithRed:0. green:1. blue:1. alpha:0.3],
                     [UIColor colorWithRed:1. green:0. blue:1. alpha:0.3]
                   ];
        
        [self addOption:@"Flip effect" selector:@selector(selectFlipEffect)];
        [self addOption:@"Float effect" selector:@selector(selectFloatEffect)];
        [self addOption:@"Fold effect" selector:@selector(selectFoldEffect)];
        [self addOption:@"Rotate effect" selector:@selector(selectRotateEffect)];
        [self addOption:@"Card effect" selector:@selector(selectCardEffect)];
        [self addOption:@"Scroll effect" selector:@selector(selectScrollEffect)];
    }
    return self;
}

- (void)viewDidLoad
{
    [super viewDidLoad];
    
    _contentView = [[UIView alloc] initWithFrame:self.exampleBounds];
    [self.view addSubview:_contentView];
    
    UIToolbar *toolbar = [[UIToolbar alloc] initWithFrame:CGRectMake(0, CGRectGetMaxY(_contentView.bounds) - 44, CGRectGetWidth(_contentView.bounds), 44)];
    toolbar.autoresizingMask = UIViewAutoresizingFlexibleWidth | UIViewAutoresizingFlexibleTopMargin;
    [_contentView addSubview:toolbar];
    
    UIBarButtonItem *buttonPrev = [[UIBarButtonItem alloc] initWithTitle:@"Prev month" style:UIBarButtonItemStylePlain target:self action:@selector(prevTouched:)];
    UIBarButtonItem *buttonNext = [[UIBarButtonItem alloc] initWithTitle:@"Next month" style:UIBarButtonItemStylePlain target:self action:@selector(nextTouched:)];
    UIBarButtonItem *space = [[UIBarButtonItem alloc] initWithBarButtonSystemItem:UIBarButtonSystemItemFlexibleSpace target:self action:nil];
    toolbar.items = @[buttonPrev, space, buttonNext];
    
    self.calendarView = [[TKCalendar alloc] initWithFrame:CGRectMake(0, 0, CGRectGetWidth(_contentView.bounds), CGRectGetHeight(_contentView.bounds) - 44)];
    self.calendarView.autoresizingMask = UIViewAutoresizingFlexibleWidth | UIViewAutoresizingFlexibleHeight;
    self.calendarView.delegate = self;
    self.calendarView.allowPinchZoom = NO;
    [_contentView addSubview:self.calendarView];
    
    TKCalendarMonthPresenter *presenter = (TKCalendarMonthPresenter*)self.calendarView.presenter;
    presenter.transitionMode = TKCalendarTransitionModeFlip;
    presenter.delegate = self;
    presenter.headerIsSticky = YES;
    presenter.contentView.backgroundColor = _colors[_colorIndex];
    _transitionMode = TKCalendarTransitionModeFlip;
}

- (void)viewDidLayoutSubviews
{
    [super viewDidLayoutSubviews];
    _contentView.frame = self.exampleBounds;
}

- (void)didReceiveMemoryWarning
{
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

- (void)prevTouched:(id)sender
{
    [_calendarView navigateBack:YES];
}

- (void)nextTouched:(id)sender
{
    [_calendarView navigateForward:YES];
}

- (void)selectFlipEffect
{
    [self setTransition:TKCalendarTransitionModeFlip isVertical:NO];
}

- (void)selectFloatEffect
{
    [self setTransition:TKCalendarTransitionModeFloat isVertical:NO];
}

- (void)selectFoldEffect
{
    [self setTransition:TKCalendarTransitionModeFold isVertical:NO];
}

- (void)selectRotateEffect
{
    [self setTransition:TKCalendarTransitionModeRotate isVertical:NO];
}

- (void)selectCardEffect
{
    [self setTransition:TKCalendarTransitionModeCard isVertical:YES];
}

- (void)selectScrollEffect
{
    [self setTransition:TKCalendarTransitionModeScroll isVertical:YES];
}

- (void)setTransition:(TKCalendarTransitionMode)transitionMode isVertical:(BOOL)isVertical
{
    _calendarView.viewMode = TKCalendarViewModeMonth;
    TKCalendarMonthPresenter *presenter = (TKCalendarMonthPresenter*)_calendarView.presenter;
    presenter.contentView.backgroundColor = _colors[_colorIndex];
    presenter.delegate = self;
    presenter.headerIsSticky = YES;
    presenter.transitionIsVertical = isVertical;
    presenter.transitionMode = transitionMode;
    _transitionMode = transitionMode;
}

#pragma mark TKCalendarDelegate

- (void)calendar:(TKCalendar *)calendar didChangedViewModeFrom:(TKCalendarViewMode)previousViewMode to:(TKCalendarViewMode)viewMode
{
    if (viewMode == TKCalendarViewModeMonth) {
        TKCalendarMonthPresenter *presenter = (TKCalendarMonthPresenter*)_calendarView.presenter;
        presenter.contentView.backgroundColor = _colors[_colorIndex];
        presenter.delegate = self;
        presenter.transitionMode = _transitionMode;
    }
}

#pragma mark TKCalendarPresenterDelegate

- (void)presenter:(id<TKCalendarPresenter>)presenter beginTransition:(TKViewTransition *)transition
{
    _oldColorIndex = _colorIndex;
    _colorIndex = (_colorIndex + 1) % _colors.count;
    TKCalendarMonthPresenter *monthPresenter = (TKCalendarMonthPresenter*)presenter;
    monthPresenter.contentView.backgroundColor = _colors[_colorIndex];
}

- (void)presenter:(id<TKCalendarPresenter>)presenter finishTransition:(BOOL)canceled
{
    if (canceled) {
        TKCalendarMonthPresenter *monthPresenter = (TKCalendarMonthPresenter*)presenter;
        monthPresenter.contentView.backgroundColor = _colors[_oldColorIndex];
        _colorIndex = _oldColorIndex;
    }
}

@end

File: TelerikUIExamplesInSwift/TelerikUIExamplesInSwift/CalendarTransitonEffects.swift
//
//  CalendarTransitonEffects.swift
//  Swift Examples
//
//  Copyright (c) 2014 Telerik. All rights reserved.
//

import Foundation

class CalendarTransitonEffects: ExampleViewController, TKCalendarPresenterDelegate, TKCalendarDelegate {
    
    let contentView = UIView()
    let toolbar = UIToolbar()
    let calendarView = TKCalendar()
    let colors = [UIColor(red: 0, green: 1, blue: 0, alpha: 0.3),
                  UIColor(red: 1, green: 0, blue: 0, alpha: 0.3),
                  UIColor(red: 0, green: 0, blue: 1, alpha: 0.3),
                  UIColor(red: 1, green: 1, blue: 0, alpha: 0.3),
                  UIColor(red: 0, green: 1, blue: 1, alpha: 0.3),
                  UIColor(red: 1, green: 0, blue: 1, alpha: 0.3)]
    var colorIndex = 0
    var oldColorIndex = -1
    var transitionMode = TKCalendarTransitionMode.Scroll

    override init(nibName nibNameOrNil: String!, bundle nibBundleOrNil: NSBundle!) {
        super.init(nibName: nibNameOrNil, bundle: nibBundleOrNil)

        self.addOption("Flip effect") { self.selectFlipEffect() }
        self.addOption("Float effect") { self.selectFloatEffect() }
        self.addOption("Fold effect") { self.selectFoldEffect() }
        self.addOption("Rotate effect") { self.selectRotateEffect() }
        self.addOption("Card effect") { self.selectCardEffect() }
        self.addOption("Scroll effect") { self.selectScrollEffect() }
    }

    required init(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        self.contentView.frame = self.exampleBounds
        self.view.addSubview(self.contentView)

        self.toolbar.frame = CGRectMake(0, CGRectGetHeight(self.contentView.bounds) - 44, CGRectGetWidth(self.contentView.bounds), 44)
        self.toolbar.autoresizingMask = UIViewAutoresizing.FlexibleWidth | UIViewAutoresizing.FlexibleTopMargin
        let buttonPrev = UIBarButtonItem(title: "Prev month", style: UIBarButtonItemStyle.Plain, target: self, action: "prevTouched:")
        let buttonNext = UIBarButtonItem(title: "Next month", style: UIBarButtonItemStyle.Plain, target: self, action: "nextTouched:")
        let space = UIBarButtonItem(barButtonSystemItem: UIBarButtonSystemItem.FlexibleSpace, target: self, action: nil)
        self.toolbar.items = [buttonPrev, space, buttonNext]
        self.contentView.addSubview(self.toolbar)
        
        self.calendarView.frame = CGRectMake(0, 0, CGRectGetWidth(self.contentView.bounds), CGRectGetHeight(self.contentView.bounds) - 44)
        self.calendarView.autoresizingMask = UIViewAutoresizing.FlexibleWidth | UIViewAutoresizing.FlexibleHeight
        self.calendarView.delegate = self
        self.calendarView.allowPinchZoom = false
        self.contentView.addSubview(self.calendarView)
        
        let presenter = self.calendarView.presenter() as! TKCalendarMonthPresenter
        presenter.transitionMode = TKCalendarTransitionMode.Flip
        presenter.delegate = self
        presenter.headerIsSticky = true
        presenter.contentView().backgroundColor = colors[colorIndex]
        self.transitionMode = TKCalendarTransitionMode.Flip
    }
    
    override func viewDidLayoutSubviews() {
        super.viewDidLayoutSubviews()
        self.contentView.frame = self.exampleBounds
    }
    
    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }
    
//MARK: - Events
    
    func prevTouched(sender: AnyObject) {
        calendarView.navigateBack(true)
    }
    
    func nextTouched(sender: AnyObject) {
        calendarView.navigateForward(true)
    }
    
    func selectFlipEffect() {
        self.setTransition(TKCalendarTransitionMode.Flip, isVertical: false)
    }
    
    func selectFloatEffect() {
        self.setTransition(TKCalendarTransitionMode.Float, isVertical: false)
    }
    
    func selectFoldEffect() {
        self.setTransition(TKCalendarTransitionMode.Fold, isVertical: false)
    }
    
    func selectRotateEffect() {
        self.setTransition(TKCalendarTransitionMode.Rotate, isVertical: false)
    }
    
    func selectCardEffect() {
        self.setTransition(TKCalendarTransitionMode.Card, isVertical: true)
    }
    
    func selectScrollEffect() {
        self.setTransition(TKCalendarTransitionMode.Scroll, isVertical: true)
    }
    
    func setTransition(transitionMode: TKCalendarTransitionMode, isVertical: Bool) {
        calendarView.viewMode = TKCalendarViewMode.Month
        let presenter = calendarView.presenter() as! TKCalendarMonthPresenter
        presenter.delegate = self
        presenter.headerIsSticky = true
        presenter.transitionIsVertical = isVertical
        presenter.transitionMode = transitionMode
        self.transitionMode = transitionMode
    }
    
//MARK: - TKCalendarDelegate
    
    func calendar(calendar: TKCalendar!, didChangedViewModeFrom previousViewMode: TKCalendarViewMode, to viewMode: TKCalendarViewMode) {
        if viewMode == TKCalendarViewMode.Month {
            let presenter = calendar.presenter() as! TKCalendarMonthPresenter
            presenter.contentView().backgroundColor = colors[colorIndex]
            presenter.delegate = self;
            presenter.transitionMode = transitionMode;
        }
    }
    
//MARK: - TKCalendarPresenterDelegate
    
    func presenter(presenter: TKCalendarPresenter!, beginTransition transition: TKViewTransition!) {
        oldColorIndex = colorIndex
        colorIndex = (colorIndex + 1) % colors.count
        let monthPresenter = presenter as! TKCalendarMonthPresenter
        monthPresenter.contentView().backgroundColor = colors[colorIndex]
    }
    
    func presenter(presenter: TKCalendarPresenter!, finishTransition canceled: Bool) {
        if canceled {
            let monthPresenter = presenter as! TKCalendarMonthPresenter
            monthPresenter.contentView().backgroundColor = colors[oldColorIndex]
            colorIndex = oldColorIndex
        }
    }
}
